File: pillar-01-iron/vantis-pqc/examples/android/VantisPQCExample/service/VantisPQCService.kt
package com.vantis.pqc.example.service

import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.security.KeyStore
import java.security.SecureRandom
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

/**
 * Service class providing a Kotlin-friendly interface to VantisPQC operations.
 * This is a placeholder that demonstrates the intended API design.
 * Actual implementation requires JNI bindings to the Rust library.
 */
class VantisPQCService {

    companion object {
        private const val KEYSTORE_PROVIDER = "AndroidKeyStore"
        private const val KEY_ALIAS_PREFIX = "vantis_pqc_"
    }

    private val keyStore: KeyStore = KeyStore.getInstance(KEYSTORE_PROVIDER).apply { load(null) }
    private var isInitialized: Boolean = false

    /**
     * Initialize the PQC service
     */
    suspend fun initialize(): Result<Unit> = withContext(Dispatchers.Default) {
        // In production, this would initialize the Rust FFI layer
        isInitialized = true
        Result.success(Unit)
    }

    // MARK: - Kyber Key Operations

    /**
     * Generate a new Kyber key pair
     * @param securityLevel The security level (Kyber512, Kyber768, Kyber1024)
     * @return A KyberKeyPairResult containing the public and private keys
     */
    suspend fun generateKyberKeyPair(
        securityLevel: KyberSecurityLevel = KyberSecurityLevel.KYBER_768
    ): Result<KyberKeyPairResult> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        // Placeholder: In production, this calls the Rust FFI
        val publicKey = generateRandomBytes(securityLevel.publicKeySize)
        val privateKey = generateRandomBytes(securityLevel.privateKeySize)

        Result.success(
            KyberKeyPairResult(
                publicKey = publicKey,
                privateKey = privateKey,
                securityLevel = securityLevel
            )
        )
    }

    /**
     * Encapsulate a shared secret using a public key
     * @param publicKey The recipient's public key
     * @param securityLevel The security level used for key generation
     * @return An encapsulation result with shared secret and ciphertext
     */
    suspend fun encapsulate(
        publicKey: ByteArray,
        securityLevel: KyberSecurityLevel = KyberSecurityLevel.KYBER_768
    ): Result<EncapsulationResult> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        // Placeholder: In production, this calls the Rust FFI
        val sharedSecret = generateRandomBytes(32)
        val ciphertext = generateRandomBytes(securityLevel.ciphertextSize)

        Result.success(
            EncapsulationResult(
                sharedSecret = sharedSecret,
                ciphertext = ciphertext
            )
        )
    }

    /**
     * Decapsulate a ciphertext using a private key
     * @param privateKey The recipient's private key
     * @param ciphertext The ciphertext to decapsulate
     * @param securityLevel The security level used for key generation
     * @return The shared secret
     */
    suspend fun decapsulate(
        privateKey: ByteArray,
        ciphertext: ByteArray,
        securityLevel: KyberSecurityLevel = KyberSecurityLevel.KYBER_768
    ): Result<ByteArray> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        if (ciphertext.size != securityLevel.ciphertextSize) {
            return@withContext Result.failure(VantisPQCError.InvalidCiphertextSize)
        }

        // Placeholder: In production, this calls the Rust FFI
        Result.success(generateRandomBytes(32))
    }

    // MARK: - Dilithium Signature Operations

    /**
     * Generate a new Dilithium key pair for signatures
     * @param securityLevel The security level (Dilithium2, Dilithium3, Dilithium5)
     * @return A DilithiumKeyPairResult containing the public and private keys
     */
    suspend fun generateDilithiumKeyPair(
        securityLevel: DilithiumSecurityLevel = DilithiumSecurityLevel.DILITHIUM_3
    ): Result<DilithiumKeyPairResult> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        // Placeholder: In production, this calls the Rust FFI
        val publicKey = generateRandomBytes(securityLevel.publicKeySize)
        val privateKey = generateRandomBytes(securityLevel.privateKeySize)

        Result.success(
            DilithiumKeyPairResult(
                publicKey = publicKey,
                privateKey = privateKey,
                securityLevel = securityLevel
            )
        )
    }

    /**
     * Sign a message using a Dilithium private key
     * @param privateKey The signer's private key
     * @param message The message to sign
     * @param securityLevel The security level used for key generation
     * @return The signature
     */
    suspend fun sign(
        privateKey: ByteArray,
        message: ByteArray,
        securityLevel: DilithiumSecurityLevel = DilithiumSecurityLevel.DILITHIUM_3
    ): Result<ByteArray> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        // Placeholder: In production, this calls the Rust FFI
        Result.success(generateRandomBytes(securityLevel.signatureSize))
    }

    /**
     * Verify a signature using a Dilithium public key
     * @param publicKey The signer's public key
     * @param message The original message
     * @param signature The signature to verify
     * @param securityLevel The security level used for key generation
     * @return True if the signature is valid
     */
    suspend fun verify(
        publicKey: ByteArray,
        message: ByteArray,
        signature: ByteArray,
        securityLevel: DilithiumSecurityLevel = DilithiumSecurityLevel.DILITHIUM_3
    ): Result<Boolean> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        // Placeholder: In production, this calls the Rust FFI
        // For demo purposes, always return true
        Result.success(true)
    }

    // MARK: - Streaming Encryption

    /**
     * Encrypt data using streaming encryption
     * @param data The data to encrypt
     * @param key The encryption key (32 bytes)
     * @param chunkSize Optional chunk size (default 64KB)
     * @return The encrypted data
     */
    suspend fun encryptStream(
        data: ByteArray,
        key: ByteArray,
        chunkSize: Int = 64 * 1024
    ): Result<ByteArray> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        if (key.size != 32) {
            return@withContext Result.failure(VantisPQCError.InvalidKeySize)
        }

        // Placeholder: In production, this uses ChaCha20-Poly1305 streaming
        // For demo, return data with overhead for nonce + auth tag
        val nonce = generateRandomBytes(12)
        val authTag = generateRandomBytes(16)
        
        Result.success(nonce + data + authTag)
    }

    /**
     * Decrypt data using streaming decryption
     * @param encryptedData The encrypted data
     * @param key The decryption key (32 bytes)
     * @return The decrypted data
     */
    suspend fun decryptStream(
        encryptedData: ByteArray,
        key: ByteArray
    ): Result<ByteArray> = withContext(Dispatchers.Default) {
        if (!isInitialized) {
            return@withContext Result.failure(VantisPQCError.NotInitialized)
        }

        if (key.size != 32) {
            return@withContext Result.failure(VantisPQCError.InvalidKeySize)
        }

        if (encryptedData.size <= 28) {
            return@withContext Result.failure(VantisPQCError.InvalidData)
        }

        // Placeholder: In production, this decrypts using ChaCha20-Poly1305
        // For demo, strip nonce and tag
        Result.success(encryptedData.copyOfRange(12, encryptedData.size - 16))
    }

    // MARK: - Android Keystore Operations

    /**
     * Store a key in the Android Keystore
     * @param key The key data
     * @param identifier A unique identifier for the key
     */
    suspend fun storeInKeystore(
        key: ByteArray,
        identifier: String
    ): Result<Unit> = withContext(Dispatchers.Default) {
        try {
            // Note: Android Keystore has limitations on key size
            // For large PQC keys, consider using encrypted file storage
            // with a Keystore-protected wrapping key
            
            val alias = "$KEY_ALIAS_PREFIX$identifier"
            
            // Create a wrapping key in the Keystore
            val keyGenerator = KeyGenerator.getInstance(
                KeyProperties.KEY_ALGORITHM_AES,
                KEYSTORE_PROVIDER
            )
            
            val spec = KeyGenParameterSpec.Builder(
                alias,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build()
            
            keyGenerator.init(spec)
            keyGenerator.generateKey()
            
            // In production, use this key to encrypt/wrap the PQC key
            // and store the wrapped key in SharedPreferences or file
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(VantisPQCError.KeystoreError(e.message ?: "Unknown error"))
        }
    }

    /**
     * Delete a key from the Android Keystore
     * @param identifier The unique identifier for the key
     */
    suspend fun deleteFromKeystore(identifier: String): Result<Unit> = withContext(Dispatchers.Default) {
        try {
            val alias = "$KEY_ALIAS_PREFIX$identifier"
            if (keyStore.containsAlias(alias)) {
                keyStore.deleteEntry(alias)
            }
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(VantisPQCError.KeystoreError(e.message ?: "Unknown error"))
        }
    }

    /**
     * Check if a key exists in the Keystore
     * @param identifier The unique identifier for the key
     */
    fun keyExists(identifier: String): Boolean {
        val alias = "$KEY_ALIAS_PREFIX$identifier"
        return keyStore.containsAlias(alias)
    }

    // MARK: - Private Helpers

    private fun generateRandomBytes(count: Int): ByteArray {
        val bytes = ByteArray(count)
        SecureRandom().nextBytes(bytes)
        return bytes
    }
}

// MARK: - Supporting Types

/**
 * Kyber security levels
 */
enum class KyberSecurityLevel(
    val publicKeySize: Int,
    val privateKeySize: Int,
    val ciphertextSize: Int,
    val nistLevel: String
) {
    KYBER_512(800, 1632, 768, "NIST Level 1"),
    KYBER_768(1184, 2400, 1088, "NIST Level 3"),
    KYBER_1024(1568, 3168, 1568, "NIST Level 5");

    val displayName: String
        get() = name.replace("_", "-")
}

/**
 * Dilithium security levels
 */
enum class DilithiumSecurityLevel(
    val publicKeySize: Int,
    val privateKeySize: Int,
    val signatureSize: Int,
    val nistLevel: String
) {
    DILITHIUM_2(1312, 2560, 2420, "NIST Level 2"),
    DILITHIUM_3(1952, 4032, 3293, "NIST Level 3"),
    DILITHIUM_5(2592, 4864, 4595, "NIST Level 5");

    val displayName: String
        get() = name.replace("_", "-")
}

/**
 * Kyber key pair result
 */
data class KyberKeyPairResult(
    val publicKey: ByteArray,
    val privateKey: ByteArray,
    val securityLevel: KyberSecurityLevel
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as KyberKeyPairResult
        return publicKey.contentEquals(other.publicKey) &&
               privateKey.contentEquals(other.privateKey) &&
               securityLevel == other.securityLevel
    }

    override fun hashCode(): Int {
        var result = publicKey.contentHashCode()
        result = 31 * result + privateKey.contentHashCode()
        result = 31 * result + securityLevel.hashCode()
        return result
    }
}

/**
 * Dilithium key pair result
 */
data class DilithiumKeyPairResult(
    val publicKey: ByteArray,
    val privateKey: ByteArray,
    val securityLevel: DilithiumSecurityLevel
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as DilithiumKeyPairResult
        return publicKey.contentEquals(other.publicKey) &&
               privateKey.contentEquals(other.privateKey) &&
               securityLevel == other.securityLevel
    }

    override fun hashCode(): Int {
        var result = publicKey.contentHashCode()
        result = 31 * result + privateKey.contentHashCode()
        result = 31 * result + securityLevel.hashCode()
        return result
    }
}

/**
 * Encapsulation result
 */
data class EncapsulationResult(
    val sharedSecret: ByteArray,
    val ciphertext: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as EncapsulationResult
        return sharedSecret.contentEquals(other.sharedSecret) &&
               ciphertext.contentEquals(other.ciphertext)
    }

    override fun hashCode(): Int {
        var result = sharedSecret.contentHashCode()
        result = 31 * result + ciphertext.contentHashCode()
        return result
    }
}

// MARK: - Errors

/**
 * VantisPQC errors
 */
sealed class VantisPQCError : Exception() {
    object NotInitialized : VantisPQCError() {
        override val message: String = "VantisPQC service is not initialized"
    }
    
    object InvalidKeySize : VantisPQCError() {
        override val message: String = "Invalid key size provided"
    }
    
    object InvalidCiphertextSize : VantisPQCError() {
        override val message: String = "Invalid ciphertext size"
    }
    
    object InvalidData : VantisPQCError() {
        override val message: String = "Invalid data provided"
    }
    
    object RandomGenerationFailed : VantisPQCError() {
        override val message: String = "Failed to generate random bytes"
    }
    
    data class KeystoreError(override val message: String) : VantisPQCError()
    
    data class Unknown(override val message: String) : VantisPQCError()
}

// MARK: - ByteArray Extensions

/**
 * Convert ByteArray to hex string
 */
fun ByteArray.toHex(): String = joinToString("") { "%02x".format(it) }

/**
 * Get first n bytes as hex string
 */
fun ByteArray.toHexPrefix(length: Int): String = take(length).toHex()

File: pillar-01-iron/vantis-pqc/examples/ios/VantisPQCExample/Services/VantisPQCService.swift
import Foundation
import Security

/// Service class providing a Swift-friendly interface to VantisPQC operations
/// This is a placeholder that demonstrates the intended API design.
/// Actual implementation requires FFI bindings to the Rust library.
public class VantisPQCService: ObservableObject {
    
    // MARK: - Published Properties
    
    @Published public var isInitialized: Bool = false
    @Published public var lastError: String?
    
    // MARK: - Initialization
    
    public init() {
        initialize()
    }
    
    /// Initialize the PQC service
    public func initialize() {
        // In production, this would initialize the Rust FFI layer
        DispatchQueue.mainAsyncAfter(deadline: .now() + 0.1) { [weak self] in
            self?.isInitialized = true
        }
    }
    
    // MARK: - Kyber Key Operations
    
    /// Generate a new Kyber key pair
    /// - Parameter securityLevel: The security level (Kyber512, Kyber768, Kyber1024)
    /// - Returns: A KyberKeyPairResult containing the public and private keys
    public func generateKyberKeyPair(securityLevel: KyberSecurityLevel = .kyber768) async throws -> KyberKeyPairResult {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        // Placeholder: In production, this calls the Rust FFI
        // let result = try await withCheckedThrowingContinuation { continuation in
        //     vantis_pqc_kyber_generate(securityLevel.rawValue) { result in
        //         continuation.resume(with: result)
        //     }
        // }
        
        let (publicKey, privateKey) = try generateRandomKeyPair(
            publicKeySize: securityLevel.publicKeySize,
            privateKeySize: securityLevel.privateKeySize
        )
        
        return KyberKeyPairResult(
            publicKey: publicKey,
            privateKey: privateKey,
            securityLevel: securityLevel
        )
    }
    
    /// Encapsulate a shared secret using a public key
    /// - Parameters:
    ///   - publicKey: The recipient's public key
    ///   - securityLevel: The security level used for key generation
    /// - Returns: An encapsulation result with shared secret and ciphertext
    public func encapsulate(publicKey: Data, securityLevel: KyberSecurityLevel = .kyber768) async throws -> EncapsulationResult {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        // Placeholder: In production, this calls the Rust FFI
        let sharedSecret = try generateRandomBytes(count: 32)
        let ciphertext = try generateRandomBytes(count: securityLevel.ciphertextSize)
        
        return EncapsulationResult(
            sharedSecret: sharedSecret,
            ciphertext: ciphertext
        )
    }
    
    /// Decapsulate a ciphertext using a private key
    /// - Parameters:
    ///   - privateKey: The recipient's private key
    ///   - ciphertext: The ciphertext to decapsulate
    ///   - securityLevel: The security level used for key generation
    /// - Returns: The shared secret
    public func decapsulate(
        privateKey: Data,
        ciphertext: Data,
        securityLevel: KyberSecurityLevel = .kyber768
    ) async throws -> Data {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        guard ciphertext.count == securityLevel.ciphertextSize else {
            throw VantisPQCError.invalidCiphertextSize
        }
        
        // Placeholder: In production, this calls the Rust FFI
        return try generateRandomBytes(count: 32)
    }
    
    // MARK: - Dilithium Signature Operations
    
    /// Generate a new Dilithium key pair for signatures
    /// - Parameter securityLevel: The security level (Dilithium2, Dilithium3, Dilithium5)
    /// - Returns: A DilithiumKeyPairResult containing the public and private keys
    public func generateDilithiumKeyPair(securityLevel: DilithiumSecurityLevel = .dilithium3) async throws -> DilithiumKeyPairResult {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        let (publicKey, privateKey) = try generateRandomKeyPair(
            publicKeySize: securityLevel.publicKeySize,
            privateKeySize: securityLevel.privateKeySize
        )
        
        return DilithiumKeyPairResult(
            publicKey: publicKey,
            privateKey: privateKey,
            securityLevel: securityLevel
        )
    }
    
    /// Sign a message using a Dilithium private key
    /// - Parameters:
    ///   - privateKey: The signer's private key
    ///   - message: The message to sign
    ///   - securityLevel: The security level used for key generation
    /// - Returns: The signature
    public func sign(
        privateKey: Data,
        message: Data,
        securityLevel: DilithiumSecurityLevel = .dilithium3
    ) async throws -> Data {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        // Placeholder: In production, this calls the Rust FFI
        return try generateRandomBytes(count: securityLevel.signatureSize)
    }
    
    /// Verify a signature using a Dilithium public key
    /// - Parameters:
    ///   - publicKey: The signer's public key
    ///   - message: The original message
    ///   - signature: The signature to verify
    ///   - securityLevel: The security level used for key generation
    /// - Returns: True if the signature is valid
    public func verify(
        publicKey: Data,
        message: Data,
        signature: Data,
        securityLevel: DilithiumSecurityLevel = .dilithium3
    ) async throws -> Bool {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        // Placeholder: In production, this calls the Rust FFI
        // For demo purposes, always return true
        return true
    }
    
    // MARK: - Streaming Encryption
    
    /// Encrypt data using streaming encryption
    /// - Parameters:
    ///   - data: The data to encrypt
    ///   - key: The encryption key (32 bytes)
    ///   - chunkSize: Optional chunk size (default 64KB)
    /// - Returns: The encrypted data
    public func encryptStream(
        data: Data,
        key: Data,
        chunkSize: Int = 64 * 1024
    ) async throws -> Data {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        guard key.count == 32 else {
            throw VantisPQCError.invalidKeySize
        }
        
        // Placeholder: In production, this uses ChaCha20-Poly1305 streaming
        // For demo, return data with overhead for auth tag
        var encrypted = try generateRandomBytes(count: 12) // nonce
        encrypted.append(data)
        encrypted.append(try generateRandomBytes(count: 16)) // auth tag
        
        return encrypted
    }
    
    /// Decrypt data using streaming decryption
    /// - Parameters:
    ///   - encryptedData: The encrypted data
    ///   - key: The decryption key (32 bytes)
    /// - Returns: The decrypted data
    public func decryptStream(
        encryptedData: Data,
        key: Data
    ) async throws -> Data {
        guard isInitialized else {
            throw VantisPQCError.notInitialized
        }
        
        guard key.count == 32 else {
            throw VantisPQCError.invalidKeySize
        }
        
        // Placeholder: In production, this decrypts using ChaCha20-Poly1305
        // For demo, strip nonce and tag
        guard encryptedData.count > 28 else {
            throw VantisPQCError.invalidData
        }
        
        return encryptedData.dropFirst(12).dropLast(16)
    }
    
    // MARK: - Keychain Operations
    
    /// Store a key in the iOS Keychain
    /// - Parameters:
    ///   - key: The key data
    ///   - identifier: A unique identifier for the key
    /// - Throws: Keychain errors if storage fails
    public func storeInKeychain(key: Data, identifier: String) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: identifier,
            kSecValueData as String: key,
            kSecAttrAccessible as String: kSecAttrAccessibleWhenUnlockedThisDeviceOnly
        ]
        
        let status = SecItemAdd(query as CFDictionary, nil)
        guard status == errSecSuccess else {
            throw VantisPQCError.keychainError(status)
        }
    }
    
    /// Retrieve a key from the iOS Keychain
    /// - Parameter identifier: The unique identifier for the key
    /// - Returns: The key data
    public func retrieveFromKeychain(identifier: String) throws -> Data {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: identifier,
            kSecReturnData as String: true
        ]
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        guard status == errSecSuccess, let data = result as? Data else {
            throw VantisPQCError.keychainError(status)
        }
        
        return data
    }
    
    /// Delete a key from the iOS Keychain
    /// - Parameter identifier: The unique identifier for the key
    public func deleteFromKeychain(identifier: String) throws {
        let query: [String: Any] = [
            kSecClass as String: kSecClassKey,
            kSecAttrApplicationTag as String: identifier
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        guard status == errSecSuccess else {
            throw VantisPQCError.keychainError(status)
        }
    }
    
    // MARK: - Private Helpers
    
    private func generateRandomBytes(count: Int) throws -> Data {
        var bytes = [UInt8](repeating: 0, count: count)
        let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
        guard status == errSecSuccess else {
            throw VantisPQCError.randomGenerationFailed
        }
        return Data(bytes)
    }
    
    private func generateRandomKeyPair(publicKeySize: Int, privateKeySize: Int) throws -> (Data, Data) {
        let publicKey = try generateRandomBytes(count: publicKeySize)
        let privateKey = try generateRandomBytes(count: privateKeySize)
        return (publicKey, privateKey)
    }
}

// MARK: - Supporting Types

/// Kyber security levels
public enum KyberSecurityLevel: Int, CaseIterable {
    case kyber512 = 1
    case kyber768 = 2
    case kyber1024 = 3
    
    public var publicKeySize: Int {
        switch self {
        case .kyber512: return 800
        case .kyber768: return 1184
        case .kyber1024: return 1568
        }
    }
    
    public var privateKeySize: Int {
        switch self {
        case .kyber512: return 1632
        case .kyber768: return 2400
        case .kyber1024: return 3168
        }
    }
    
    public var ciphertextSize: Int {
        switch self {
        case .kyber512: return 768
        case .kyber768: return 1088
        case .kyber1024: return 1568
        }
    }
    
    public var nistLevel: String {
        switch self {
        case .kyber512: return "NIST Level 1"
        case .kyber768: return "NIST Level 3"
        case .kyber1024: return "NIST Level 5"
        }
    }
}

/// Dilithium security levels
public enum DilithiumSecurityLevel: Int, CaseIterable {
    case dilithium2 = 1
    case dilithium3 = 2
    case dilithium5 = 3
    
    public var publicKeySize: Int {
        switch self {
        case .dilithium2: return 1312
        case .dilithium3: return 1952
        case .dilithium5: return 2592
        }
    }
    
    public var privateKeySize: Int {
        switch self {
        case .dilithium2: return 2560
        case .dilithium3: return 4032
        case .dilithium5: return 4864
        }
    }
    
    public var signatureSize: Int {
        switch self {
        case .dilithium2: return 2420
        case .dilithium3: return 3293
        case .dilithium5: return 4595
        }
    }
    
    public var nistLevel: String {
        switch self {
        case .dilithium2: return "NIST Level 2"
        case .dilithium3: return "NIST Level 3"
        case .dilithium5: return "NIST Level 5"
        }
    }
}

/// Kyber key pair result
public struct KyberKeyPairResult: Codable {
    public let publicKey: Data
    public let privateKey: Data
    public let securityLevel: KyberSecurityLevel
}

/// Dilithium key pair result
public struct DilithiumKeyPairResult: Codable {
    public let publicKey: Data
    public let privateKey: Data
    public let securityLevel: DilithiumSecurityLevel
}

/// Encapsulation result
public struct EncapsulationResult: Codable {
    public let sharedSecret: Data
    public let ciphertext: Data
}

// MARK: - Errors

/// VantisPQC errors
public enum VantisPQCError: Error, LocalizedError {
    case notInitialized
    case invalidKeySize
    case invalidCiphertextSize
    case invalidData
    case randomGenerationFailed
    case keychainError(OSStatus)
    case encodingError
    case unknown(String)
    
    public var errorDescription: String? {
        switch self {
        case .notInitialized:
            return "VantisPQC service is not initialized"
        case .invalidKeySize:
            return "Invalid key size provided"
        case .invalidCiphertextSize:
            return "Invalid ciphertext size"
        case .invalidData:
            return "Invalid data provided"
        case .randomGenerationFailed:
            return "Failed to generate random bytes"
        case .keychainError(let status):
            return "Keychain error: \(status)"
        case .encodingError:
            return "Encoding/decoding error"
        case .unknown(let message):
            return message
        }
    }
}